### PR TITLE
Adding a behavior to add MessageType header when missing

### PR DESCRIPTION
**Pull request by WojcikMike**
_Connects to https://github.com/Particular/NServiceBus/issues/4360_

### CODE REVIEW

**Comment by HEskandari** [comment]
@mikeminutillo this is ready for review as the build is green. LMK if I've missed anything.

File: src/NServiceBus.Core/Pipeline/Incoming/MessageTypeEnricherBehavior.cs [hunk 1]
```diff
@@ -0,0 +1,19 @@
+namespace NServiceBus
+{
+    using System;
+    using System.Threading.Tasks;
+    using Pipeline;
+
+    class MessageTypeEnricherBehavior : Behavior<IIncomingLogicalMessageContext>
```
Comment (mikeminutillo) [inline src/NServiceBus.Core/Pipeline/Incoming/MessageTypeEnricherBehavior.cs:7]: Should the name (of the behavior and the associated feature) be named more explicitly? There is already an existing behavior that is responsible for setting the message type header on outgoing messages. Should this one be named something like `InferredMessageTypeEnricher` or something?

Comment (HEskandari) [reply src/NServiceBus.Core/Pipeline/Incoming/MessageTypeEnricherBehavior.cs:7]: Done

File: src/NServiceBus.Core/Pipeline/Incoming/MessageTypeEnricherBehavior.cs [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+    {
+        public override async Task Invoke(IIncomingLogicalMessageContext context, Func<Task> next)
```
[the remark this answers is not in the document]
Comment (HEskandari) [reply src/NServiceBus.Core/Pipeline/Incoming/MessageTypeEnricherBehavior.cs:9]: Done.

File: src/NServiceBus.Core/Routing/MessageTypeEnricherFeature.cs [hunk 3]
```diff
@@ -0,0 +1,15 @@
...
+    class MessageTypeEnricherFeature : Feature
+    {
+        public MessageTypeEnricherFeature()
+        {
+            EnableByDefault();
+        }
+
+        protected internal override void Setup(FeatureConfigurationContext context)
+        {
+            context.Pipeline.Register(typeof(MessageTypeEnricherBehavior), "Adds EnclosedMessageType to the header of the message if it doesn't exist.");
```
[the remark this answers is not in the document]
Comment (HEskandari) [reply src/NServiceBus.Core/Routing/MessageTypeEnricherFeature.cs]: Updated

**Pull request merged**